File: Lab2.4/С#/Expression.cs
using System;

namespace MathLibrary
{
    public class NegativeLogarithmException : Exception
    {
        public NegativeLogarithmException()
            : base("Attempt to calculate logarithm of non-positive number") { }
    }

    public class Expression
    {
        private double a;
        private double b;
        private double c;

        public Expression(double a, double b, double c)
        {
            this.a = a;
            this.b = b;
            this.c = c;
        }

        public double A
        {
            get { return a; }
            set { a = value; }
        }

        public double B
        {
            get { return b; }
            set { b = value; }
        }

        public double C
        {
            get { return c; }
            set { c = value; }
        }

        public double CalculateLogarithm(double value)
        {
            if (value <= 0)
            {
                throw new NegativeLogarithmException();
            }
            return Math.Log(value);
        }

        public double Calculate()
        {
            try
            {
                double logResult = CalculateLogarithm(a);
                return (logResult + b) / c;
            }
            catch (NegativeLogarithmException)
            {
                throw; // Перебрасываем исключение дальше
            }
        }
    }
}

File: Lab2.4/C++/Expression.cpp
#include "Expression.h"

Expression::Expression(double a, double b, double c) : a(a), b(b), c(c) {}

double Expression::getA() const { return a; }
double Expression::getB() const { return b; }
double Expression::getC() const { return c; }

void Expression::setA(double value) { a = value; }
void Expression::setB(double value) { b = value; }
void Expression::setC(double value) { c = value; }

double Expression::calculateLogarithm(double value) const {
    if (value <= 0) {
        throw NegativeLogarithmException();
    }
    return log(value);
}

double Expression::calculate() const {
    try {
        double logResult = calculateLogarithm(a);
        return (logResult + b) / c;
    }
    catch (const NegativeLogarithmException& e) {
        throw; // Перебрасываем исключение дальше
    }
}